nemo/collections/nlp/modules/common/transformer/transformer_generators.py

```diff
@@ -372,9 +381,355 @@ def _forward(self, decoder_input_ids=None, encoder_hidden_states=None, encoder_i
+            if pad_profile.sum() == batch_size * self.beam_size:
+                break
+
+        # select best performing hypotheses in each element of the batch
+        len_penalties = self.compute_len_penalty(prefixes_len, self.len_pen)
+        scores = scores / len_penalties
+        best_guesses = (
+            torch.argmax(scores.view(-1, self.beam_size), dim=1, keepdim=True).repeat(1, prefixes.size(1)).unsqueeze(1)
+        )
+        tgt = prefixes.view(batch_size, self.beam_size, -1).gather(1, best_guesses).squeeze(1)
+
+        if return_beam_scores:
+            return prefixes, scores, tgt
+        else:
+            return tgt
+
+    def __call__(self, src_ids, encoder_input_mask, decoder_input_ids=None, return_beam_scores=False):
+        with self.as_frozen():
+            return self._forward(src_ids, encoder_input_mask, decoder_input_ids, return_beam_scores)
+
+    def freeze(self) -> None:
+        """Freeze weights of embedding, decoder, and classification layers to prevent memory leak.
+        """
+        for model_num in range(self.num_models):
+            for param in self.embeddings[model_num].parameters():
+                param.requires_grad = False
+            self.embeddings[model_num].eval()
+            for param in self.decoders[model_num].parameters():
+                param.requires_grad = False
+            self.decoders[model_num].eval()
+            for param in self.log_softmaxes[model_num].parameters():
+                param.require_grad = False
+            self.log_softmaxes[model_num].eval()
+            for param in self.encoders[model_num].parameters():
+                param.require_grad = False
+            self.encoders[model_num].eval()
+
+    def unfreeze(self) -> None:
+        """Unfreeze weights of embedding, decoder, and classification layers.
+        """
+        for model_num in range(self.num_models):
+            for param in self.embeddings[model_num].parameters():
+                param.requires_grad = True
+            self.embeddings[model_num].train()
+            for param in self.decoders[model_num].parameters():
+                param.requires_grad = True
+            self.decoders[model_num].train()
+            for param in self.log_softmaxes[model_num].parameters():
+                param.require_grad = True
+            self.log_softmaxes[model_num].train()
+            for param in self.encoders[model_num].parameters():
+                param.require_grad = True
+            self.encoders[model_num].train()
+
+    @contextmanager
+    def as_frozen(self):
+        """
+        Context manager which temporarily freezes embedding, decoder, and log_softmax modules,
+        yields control and finally unfreezes the modules.
+        """
+        self.freeze()
 
-        return tgt.squeeze(1)
+        try:
+            yield
+        finally:
+            self.unfreeze()
 
 
 class BeamSearchSequenceGeneratorWithLanguageModel(GreedySequenceGenerator):
```

Comment:
It seems that EnsembleBeamSearchSequenceGenerator has all the functional provided by BSSGWithLanguageModel. Thus, we can deprecate the latter.

Reply:
Cool, I can move the BSSGWithLanguageModel logic into BeamSearchSequenceGenerator. Does that work?

Reply:
I would rather remove it completely and leave BeamSearchSequenceGenerator as it is. Currently it implements clean Beam Search logic for a single model without extra models overhead. Let's have all advanced stuff in EBSSG.

Reply:
Okay sounds good